runtime/fiber.c

```diff
@@ -0,0 +1,534 @@
+    if (stack == NULL) {
+      return NULL;
+    }
+
+    stack->size_bucket = size_bucket;
+
+    /* Ensure 16-byte alignment because some architectures require it */
+    hand = (struct stack_handler*)
+     (((uintnat)stack + sizeof(struct stack_info) + sizeof(value) * wosize + 8)
+      & ((uintnat)-1 << 4));
+    stack->handler = hand;
+  }
+
+  hand->handle_value = hval;
+  hand->handle_exn = hexn;
+  hand->handle_effect = heff;
+  hand->parent = NULL;
+  stack->sp = (value*)hand;
+  stack->exception_ptr = NULL;
+#ifdef DEBUG
+  stack->magic = 42;
+#endif
+  CAMLassert(Stack_high(stack) - Stack_base(stack) == wosize ||
+             Stack_high(stack) - Stack_base(stack) == wosize + 1);
+  return stack;
+
+}
+
+/* allocate a stack with at least "wosize" usable words of stack */
+static struct stack_info* alloc_stack_noexc(mlsize_t wosize, value hval,
+                                            value hexn, value heff)
+{
+  struct stack_info** size_bucket = stack_cache_bucket (wosize);
+  return alloc_size_class_stack_noexc(wosize, size_bucket, hval, hexn, heff);
+}
+
+#ifdef NATIVE_CODE
+
+value caml_alloc_stack (value hval, value hexn, value heff) {
+  struct stack_info* stack =
+    alloc_size_class_stack_noexc(caml_fiber_wsz, Caml_state->stack_cache,
+                                 hval, hexn, heff);
+
+  if (!stack) caml_raise_out_of_memory();
+
+  fiber_debug_log ("Allocate stack=%p of %lu words", stack, caml_fiber_wsz);
+
+  return Val_ptr(stack);
+}
+
+void caml_get_stack_sp_pc (struct stack_info* stack,
+                           char** sp /* out */, uintnat* pc /* out */)
+{
+  char* p = (char*)stack->sp;
+
+  p += sizeof(value);
+  *sp = p;
+  *pc = Saved_return_address(*sp);
+}
+
+Caml_inline void scan_stack_frames(scanning_action f, void* fdata,
+                                   struct stack_info* stack, value* gc_regs)
+{
+  char * sp;
+  uintnat retaddr;
+  value * regs;
+  frame_descr * d;
+  uintnat h;
+  int n, ofs;
+  unsigned short * p;
+  value *root;
+  caml_frame_descrs fds = caml_get_frame_descrs();
+
+  sp = (char*)stack->sp;
+  regs = gc_regs;
+
+next_chunk:
+  if (sp == (char*)Stack_high(stack)) return;
+
+  retaddr = *(uintnat*)sp;
+  sp += sizeof(value);
+
+  while(1) {
+    /* Find the descriptor corresponding to the return address */
+    h = Hash_retaddr(retaddr, fds.mask);
+    while(1) {
+      d = fds.descriptors[h];
+      if (d->retaddr == retaddr) break;
+      h = (h+1) & fds.mask;
+    }
+    if (d->frame_size != 0xFFFF) {
+      /* Scan the roots in this frame */
+      for (p = d->live_ofs, n = d->num_live; n > 0; n--, p++) {
+        ofs = *p;
+        if (ofs & 1) {
+          root = regs + (ofs >> 1);
+        } else {
+          root = (value *)(sp + ofs);
+        }
+        f (fdata, *root, root);
+      }
+      /* Move to next frame */
+      sp += (d->frame_size & 0xFFFC);
+      retaddr = Saved_return_address(sp);
+      /* XXX KC: disabled already scanned optimization. */
```

Comment:
unnecessary comment now?

Reply:
This optimisation is only available on POWER. https://github.com/ocaml/ocaml/blob/cce52acc7c7903e92078e9fe40745e11a1b944f0/runtime/caml/stack.h#L47. The note may be useful when we implement POWER backend.

Reply:
> This optimisation is only available on POWER. 
And it doesn't seem to make much of a difference in GC time, so we may want to remove it entirely in OCaml 5.